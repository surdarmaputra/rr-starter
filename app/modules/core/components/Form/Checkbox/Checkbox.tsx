import * as CheckboxPrimitive from '@radix-ui/react-checkbox';
import { Check } from 'lucide-react';
import * as React from 'react';

import { mergeClassNames } from '~/modules/core/libs/utils/ui';

export type CheckboxProps = React.ComponentPropsWithoutRef<
  typeof CheckboxPrimitive.Root
>;

const Checkbox = React.forwardRef<
  React.ComponentRef<typeof CheckboxPrimitive.Root>,
  CheckboxProps
>(({ className, ...props }, ref) => (
  <CheckboxPrimitive.Root
    className={mergeClassNames(
      'peer h-4 w-4 shrink-0 rounded-sm border border-secondary ring-offset-white',
      'focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-primary',
      'focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50',
      'data-[state=checked]:border-primary data-[state=checked]:bg-primary data-[state=checked]:text-white',
      className,
    )}
    ref={ref}
    {...props}
  >
    <CheckboxPrimitive.Indicator
      className={mergeClassNames(
        'flex items-center justify-center text-current',
      )}
    >
      <Check className="h-4 w-4" />
    </CheckboxPrimitive.Indicator>
  </CheckboxPrimitive.Root>
));
Checkbox.displayName = CheckboxPrimitive.Root.displayName;

export { Checkbox };
