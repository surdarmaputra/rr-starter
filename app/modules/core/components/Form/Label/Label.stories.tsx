import type { Meta, StoryObj } from '@storybook/react';

import { Label, type LabelProps } from './Label';

type Story = StoryObj<typeof Label>;

const meta: Meta = {
  component: Label,
  title: 'Core/Components/Form/Label',
};

export default meta;

export const Default: Story = {
  render: (args: LabelProps) => <Label {...args}>Sample label</Label>,
};
