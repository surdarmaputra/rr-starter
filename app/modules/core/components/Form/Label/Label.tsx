import * as LabelPrimitive from '@radix-ui/react-label';
import { cva, type VariantProps } from 'class-variance-authority';
import * as React from 'react';

import { cn } from '~/modules/core/libs/shadcn/utils';

const labelVariants = cva(
  'text-sm leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70 text-letter-body',
);

export interface LabelProps
  extends React.ComponentPropsWithoutRef<typeof LabelPrimitive.Root>,
    VariantProps<typeof labelVariants> {
  secondaryText?: React.ReactNode;
}

const Label = React.forwardRef<
  React.ComponentRef<typeof LabelPrimitive.Root>,
  LabelProps
>(({ children, className, secondaryText, ...props }, ref) => (
  <LabelPrimitive.Root
    className={cn(labelVariants(), className)}
    ref={ref}
    {...props}
  >
    {children}
    {Boolean(secondaryText) && (
      <span className="ml-1.5 font-normal text-letter-caption">
        {secondaryText}
      </span>
    )}
  </LabelPrimitive.Root>
));
Label.displayName = LabelPrimitive.Root.displayName;

export { Label };
