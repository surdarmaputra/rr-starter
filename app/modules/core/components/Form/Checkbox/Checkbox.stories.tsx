import type { Meta, StoryObj } from '@storybook/react';

import { Label } from '../Label/Label';
import { Checkbox, type CheckboxProps } from './Checkbox';

type Story = StoryObj<typeof Checkbox>;

const meta: Meta = {
  component: Checkbox,
  title: 'Core/Components/Form/Checkbox',
};

export default meta;

export const Default: Story = {
  render: (args: CheckboxProps) => (
    <div className="flex items-center space-x-2">
      <Checkbox id="terms" {...args} />
      <Label htmlFor="terms">Accept terms and conditions</Label>
    </div>
  ),
};
