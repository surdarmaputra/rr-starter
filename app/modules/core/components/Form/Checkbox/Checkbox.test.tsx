import type { CheckboxProps } from '@radix-ui/react-checkbox';
import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import { describe, expect, test, vi } from 'vitest';

import { Label } from '../Label/Label';
import { Checkbox } from './Checkbox';

describe('Checkbox', () => {
  function renderComponent(props?: CheckboxProps) {
    return render(
      <div className="flex items-center space-x-2">
        <Checkbox id="terms" {...props} />
        <Label htmlFor="terms">Accept terms and conditions</Label>
      </div>,
    );
  }

  test('triggers onCheckedChange', async () => {
    const handleCheckedChange = vi.fn();
    renderComponent({ onCheckedChange: handleCheckedChange });

    const checkbox = screen.getByRole('checkbox', {
      name: /Accept terms and conditions/,
    });
    fireEvent.click(checkbox);

    await waitFor(() => expect(handleCheckedChange).toHaveBeenCalledTimes(1));
    expect(handleCheckedChange).toHaveBeenCalledWith(true);
  });

  test('renders snapshot', async () => {
    const { container } = renderComponent();
    expect(container).toMatchSnapshot();
  });
});
