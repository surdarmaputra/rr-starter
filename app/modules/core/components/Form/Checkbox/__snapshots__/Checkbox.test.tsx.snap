// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Checkbox > renders snapshot 1`] = `
<div>
  <div
    class="flex items-center space-x-2"
  >
    <button
      aria-checked="false"
      class="peer h-4 w-4 shrink-0 rounded-sm border order-slate-900 ring-offset-white focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-slate-950 focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50 data-[state=checked]:bg-slate-900 data-[state=checked]:text-slate-50 dark:border-slate-50 dark:ring-offset-slate-950 dark:focus-visible:ring-slate-300 dark:data-[state=checked]:bg-slate-50 dark:data-[state=checked]:text-slate-900"
      data-state="unchecked"
      id="terms"
      role="checkbox"
      type="button"
      value="on"
    />
    <label
      class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
      for="terms"
    >
      Accept terms and conditions
    </label>
  </div>
</div>
`;
